### PR TITLE
Update Eval API

### PR DESCRIPTION
**Pull request by hanlint**
Based on discussions in #1429 , it's clear that our Eval API should be updated to be consistent with the Trainer API, eg.
```
trainer = Trainer(
    eval_dataloader=...
)
trainer.eval()
```
or
```
trainer = Trainer(
    ...
)

trainer.eval(
    eval_dataloader=...
)
```

Instead, currently the `eval` API is fairly constrained to evaluating a single dataloader/metric, and also requires re-passing in the metrics, even though they have already been defined by the user in the `model`. This PR implements the new Evaluation API.

Todo:
- [x] update docstrings / docs
- [x] update tests

### CODE REVIEW

File: composer/trainer/trainer.py [hunk 1]
```diff
@@ -978,7 +978,7 @@ def __init__(
         if eval_dataloader is None:
             evaluators: List[Evaluator] = []
         else:
-            eval_metrics = self.state.model.get_metrics(is_train=False)
+            eval_metrics = deepcopy(self.state.model.get_metrics(is_train=False))
```
Comment (A-Jacobson) [inline composer/trainer/trainer.py:981]: Doing this would make it so that the metrics on the model aren't updated. is this intended?

Comment (hanlint) [reply composer/trainer/trainer.py:981]: Yes, in the case of multiple metrics, since the model metrics only returns `Accuracy()`, we need to duplicate them.

**Review by A-Jacobson — APPROVED** [review]
LGTM, left note about copying metrics. copying them feels strange to me since the model parameters are updated in place + if the metrics belong and are meant to be accessed from the trainer.. why are they defined on the model?

File: composer/trainer/trainer.py [hunk 2]
```diff
@@ -2140,12 +2140,155 @@ def predict_batch_end(self, state: State, logger: Logger) -> None:
...
 
     def eval(
+        self,
+        eval_dataloader: Optional[Union[Iterable, DataSpec, Evaluator, Sequence[Evaluator]]] = None,
+        subset_num_batches: int = -1,
+        # deprecated arguments below
+        dataloader: Optional[Any] = None,
+        dataloader_label: Optional[Any] = None,
+        metrics: Optional[Any] = None,
+        log_level: Optional[Any] = None,
```
Comment (eracah) [inline composer/trainer/trainer.py:2150]:
```suggestion
        self,
        eval_dataloader: Optional[Union[Iterable, DataSpec, Evaluator, Sequence[Evaluator]]] = None,
        subset_num_batches: int = -1,
        # deprecated arguments below
        **kwargs,
```
How about we just set this to kwargs and then we check if any of the deprecated arguments are specified inside the kwargs? This way we don't have args in the signature that we don't want people to use, which would be confusing.

File: composer/trainer/trainer.py [hunk 3]
```diff
@@ -2140,12 +2140,155 @@ def predict_batch_end(self, state: State, logger: Logger) -> None:
...
+                ``eval_dataloader`` provided to the trainer init().
+            subset_num_batches (int, optional): Evaluate on this many batches. Default to ``-1`` (the entire
+                dataloader. Can also be provided in the trainer init()as ``eval_subset_num_batches``.
+            dataloader: Removed, do not use.
+            dataloader_label: Removed, do not use.
+            metrics: Removed, do not use.
+            log_level: Removed, do not use.
+
+        """
+        if any([dataloader, dataloader_label, metrics]):
```
Comment (eracah) [inline composer/trainer/trainer.py:2239]:
```suggestion
        if any([k in kwargs for k in ['dataloader', 'dataloader_label', 'metrics', 'log_level']]):
```
see above comment for context. Basically, we have a catch-all kwargs that we check for deprecated arguments.

**Review by eracah — APPROVED** [review]
Looks pretty good, but maybe using a catch-all kwargs instead of deprecated arguments is easier?

File: composer/trainer/trainer.py (hunk 2 above)
Comment (hanlint) [reply composer/trainer/trainer.py:2150]: Ah i like this better, thanks!
